experiment/simulation/js/main.js

```diff
@@ -1 +1,579 @@
+let dp = [];
+// fix the heights of the canvas
+let slope = 0;
+let currx, curry;
+let dx, dy;
+let last_move_direction = "";
+let valid = 1;
+
+// let submit_button = document.getElementById("submit");
+// store the number of times next is called
+let times_next_called = 0;
+
+formx1.onchange = function () {
+  let x1 = formx1.value;
+  if (x1 < -45 || x1 > 44) {
+    valid = 0;
+    formx1.style.border = "1px solid red";
+    errorx1.innerHTML = "x should be between " + "-45" + " and " + "44";
+  } else {
+    valid = 1;
+    formx1.style.border = "none";
+    errorx1.innerHTML = "";
+  }
+};
+formy1.onchange = function () {
+  let y1 = formy1.value;
+  if (y1 < -23 || y1 > 21) {
+    valid = 0;
+    formy1.style.border = "1px solid red";
+    errory1.innerHTML = "y should be between " + "-23" + " and " + "21";
+  } else {
+    valid = 1;
+    formy1.style.border = "none";
+
+    errory1.innerHTML = "";
+  }
+};
+formx2.onchange = function () {
+  let x2 = formx2.value;
+  if (x2 < -45 || x2 > 44) {
+    formx2.style.border = "1px solid red";
+    valid = 0;
+    errorx2.innerHTML = "x should be between " + "-45" + " and " + "44";
+  } else {
+    valid = 1;
+    formx2.style.border = "none";
+   
+    errorx2.innerHTML = "";
+  }
+};
+formy2.onchange = function () {
+  let y2 = formy2.value;
+  if (y2 < -23 || y2 > 21) {
+    formy2.style.border = "1px solid red";
+    valid = 0;
+    errory2.innerHTML = "y should be between " + "-23" + " and " + "21";
+  } else {
+    valid = 1;
+    formy2.style.border = "none";
+  
+    errory2.innerHTML = "";
+  }
+};
+function makeAxis() {
+  let maxx = (MAXX / 2 + 1 + 0.5) * block_size;
+  let maxy = height - (MAXY / 2 + 1 + 0.5) * block_size;
+  let miny = height - (MINY - 1 + 0.5) * block_size;
+  let mxx = (MAXX + 1 + 0.5) * block_size;
+  let mnx = (MINX - 1 + 0.5) * block_size;
+  let mxy = height - (MAXY + 1 + 0.5) * block_size;
+  ctx.beginPath();
+  ctx.lineWidth = 2;
+  ctx.strokeStyle = "red";
+  ctx.moveTo(maxx, miny);
+  ctx.lineTo(maxx, mxy);
+  ctx.stroke();
+  ctx.beginPath();
+  ctx.lineWidth = 2;
+  ctx.strokeStyle = "red";
+  ctx.moveTo(mnx, maxy);
+  ctx.lineTo(mxx, maxy);
+  ctx.stroke();
+}
+function setParameters() {
+  // get  & set the coordinates from the input form
+  let x1 = Number(document.getElementById("x1").value);
+  let y1 = Number(document.getElementById("y1").value);
+  let x2 = Number(document.getElementById("x2").value);
+  let y2 = Number(document.getElementById("y2").value);
+
+  x1 = x1 + Number(originx);
+  y1 = y1 + Number(originy);
+  x2 = x2 + Number(originx);
+  y2 = y2 + Number(originy);
+
+  point1 = [x1, y1];
+  point2 = [x2, y2];
+  flag = 1;
+  dy = Math.abs(y2 - y1);
+  dx = Math.abs(x2 - x1);
+  currx = point1[0];
+  curry = point1[1];
+  if (point1[0] == point2[0]) {
+    if (point1[1] > point2[1]) slope = Number.MIN_SAFE_INTEGER;
+    else slope = Number.MAX_SAFE_INTEGER;
+  } else {
+    slope = (point2[1] - point1[1]) / (point2[0] - point1[0]);
```

Comment:
It may be a good idea to declare a separate function computeLineSlope and take all this calculation there.